This week, we learned about ensemble methods, APIs, and natural language processing. Now we're going to put these skills to the test. You've been hired by Netflix to examine what factors lead to certain ratings for movies. Given that Netflix does not currently store this type of data, your boss has suggested that you collect ratings and reviews data from IMDB. Netflix is no stranger to machine learning, however:

Netflix uses random forests and decision trees to predict what types of movies an individual user may like.
Using unsupervised learning techniques, they are able to continually update suggestions, listings, and other features of it's user interface.
Netflix, however, hasn't focused on collecting data on the top movies of all time, and would like to add some of them to their offerings based on popularity and other factors.

Point: Your boss isn't sure where to start on this project, so your task is to collect the data and construct a random forest to understand what factors contribute to ratings.


Acquire data from IMDB, and use whatever metrics you can collect to predict whether it is a good movie.

When you've finished your analysis, Netflix would like a report detailing your findings, with recommendations as to next steps.

Here are some questions to keep in mind:

What factors are the most direct predictors of rating?
You can use rating as your target variable. But it's up to you whether to treat it as continuous, binary, or multiclass.
Goal: Completed Jupyter notebook that includes modeling using a random forest and an blog post explaining your findings.

This is deliberately open ended. There is no starter code. It's up to you how to acquire the data, store the data, and what features you want to use.

We expect you to use a tree-based model, but the rest of the decisions are up to you.

We will be looking for the following things:

A clear problem statement & description of the goals of your study to be included in the final report

Data from IMDB

Cleaned and refined data

Visualization. Plots that describe your data and evaluate your model.

Tree-based models (use any combination of ensemble techniques: random forests, bagging, boosting).

A blog post presenting the results of your findings as a report to Netflix, including:
a problem statement, summary statistics of the various factors (e.g. year, number of ratings, etc.), your model, at least 2 graphics, and your recommendations for next steps!

In [ ]:
>>> imdb.top_250()
[{'title': 'The Shawshank Redemption', 'year': '1994', 'type': 'feature', 'rating': 9.3,...}, ...]


You can get data on the top 250 movies on IMDB using the IMBDpie API
If you need additional data, you can either research additional APIs, or scrape it yourself using BeautifulSoup
Read the docs for whatever technologies you use. Most of the time, there is a tutorial that you can follow, but not always, and learning to read documentation is crucial to your success!
Document everything.

pseudo-code for part one of project:
    -webscrape top 250 list from IMDB using IMDBpie API
    -plug in list of top 250 movies in OMDB API to pull other needed data
    -compile database of top 250 movies and relevant data that can be obtained
    -clean DB and get ready for decision-tree analysis.....

tonight's to-do list:

1) Get far along project:

    goal: 
    a) Data from IMDB
    b) Cleaned and refined data
    c) Visualization. Plots that describe your data
    
2) review ensemble methods, APIs, and natural language processing
    write pseudo code for this

In [94]:
import requests
import bs4
from bs4 import BeautifulSoup, BeautifulStoneSoup
import urllib2
# from IPython.display import HTML
import re
import numpy as np
import pandas as pd

In [83]:
from imdbpie import Imdb
imdb = Imdb()
imdb = Imdb(anonymize=True) # to proxy requests

# Creating an instance with caching enabled
# Note that the cached responses expire every 2 hours or so.
# The API response itself dictates the expiry time)
imdb = Imdb(cache=True)

In [84]:
top250 = imdb.top_250()

In [85]:
import pandas as pd

idf = pd.DataFrame(top250)
idf.tail()

can_rate                                              image  num_votes  \
245     True  {u'url': u'http://ia.media-imdb.com/images/M/M...      31510   
246     True  {u'url': u'http://ia.media-imdb.com/images/M/M...      79767   
247     True  {u'url': u'http://ia.media-imdb.com/images/M/M...      44750   
248     True  {u'url': u'http://ia.media-imdb.com/images/M/M...     306971   
249     True  {u'url': u'http://ia.media-imdb.com/images/M/M...      75209   

     rating     tconst                 title     type  year  
245     8.1  tt0050613       Throne of Blood  feature  1957  
246     8.1  tt0044706             High Noon  feature  1952  
247     8.3  tt1954470    Gangs of Wasseypur  feature  2012  
248     8.0  tt0101414  Beauty and the Beast  feature  1991  
249     8.1  tt0118694  In the Mood for Love  feature  2000

using 250 list, input title into URL for OMDB API thing for every title, and put result data into dataframe.
Once I have 250 data rows, put together with first DF (if necessary)....may not be necessary


In [86]:
results = []


for title in idf['title']:
    title = title.encode('utf8')
    title = title.replace(" ", "+")
    URL = "http://www.omdbapi.com/?t={title}&plot=full&r=xml&tomatoes=true".format(**locals())
    i = requests.get(URL)
    i_soup = BeautifulSoup(i.content, "lxml-xml")
    results.append(i_soup)
    
results

[<?xml version="1.0" encoding="unicode-escape"?>\n<root response="True"><movie BoxOffice="N/A" DVD="27 Jan 1998" Production="Columbia Pictures" Website="N/A" actors="Tim Robbins, Morgan Freeman, Bob Gunton, William Sadler" awards="Nominated for 7 Oscars. Another 15 wins &amp; 21 nominations." country="USA" director="Frank Darabont" genre="Crime, Drama" imdbID="tt0111161" imdbRating="9.3" imdbVotes="1,668,622" language="English" metascore="80" plot="Chronicles the experiences of a formerly successful banker as a prisoner in the gloomy jailhouse of Shawshank after being found guilty of a crime he did not commit. The film portrays the man's unique way of dealing with his new, torturous life; along the way he befriends a number of fellow prisoners, most notably a wise long-term inmate named Red." poster="http://ia.media-imdb.com/images/M/MV5BODU4MjU4NjIwNl5BMl5BanBnXkFtZTgwMDU2MjEyMDE@._V1_SX300.jpg" rated="R" released="14 Oct 1994" runtime="142 min" title="The Shawshank Redemption" tomato

In [66]:
xml = "<tag>Text 1<selfclosing>Text 2"
for soups in results:
    print type(soups)
#     print BeautifulStoneSoup(soups).prettify()

<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.BeautifulSoup'>
<class 'bs4.Beautifu

[<?xml version="1.0" encoding="UTF-8"?><html><body><root response="True"><movie actors="Tim Robbins, Morgan Freeman, Bob Gunton, William Sadler" awards="Nominated for 7 Oscars. Another 15 wins &amp; 21 nominations." boxoffice="N/A" country="USA" director="Frank Darabont" dvd="27 Jan 1998" genre="Crime, Drama" imdbid="tt0111161" imdbrating="9.3" imdbvotes="1,668,622" language="English" metascore="80" plot="Chronicles the experiences of a formerly successful banker as a prisoner in the gloomy jailhouse of Shawshank after being found guilty of a crime he did not commit. The film portrays the man's unique way of dealing with his new, torturous life; along the way he befriends a number of fellow prisoners, most notably a wise long-term inmate named Red." poster="http://ia.media-imdb.com/images/M/MV5BODU4MjU4NjIwNl5BMl5BanBnXkFtZTgwMDU2MjEyMDE@._V1_SX300.jpg" production="Columbia Pictures" rated="R" released="14 Oct 1994" runtime="142 min" title="The Shawshank Redemption" tomatoconsensus="The Shawshank Redemption is an uplifting, deeply satisfying prison drama with sensitive direction and fine performances." tomatoeurl="http://www.rottentomatoes.com/m/shawshank_redemption/" tomatofresh="60" tomatoimage="certified" tomatometer="91" tomatorating="8.2" tomatoreviews="66" tomatorotten="6" tomatousermeter="98" tomatouserrating="4.4" tomatouserreviews="881281" type="movie" website="N/A" writer="Stephen King (short story &amp;quot;Rita Hayworth and Shawshank Redemption&amp;quot;), Frank Darabont (screenplay)" year="1994"></movie></root></body></html>

In [112]:
def extract_everything(list_of_soup_files):
    """Extract desired movie info using BS4. Future: will refactor using modulo (%) 
    to pass in a list of desired tag/attribute names if possible..."""
    list_dict_250 = []
    for soup_file in list_of_soup_files:
        dict_250 = {}
        
        ratings=soup_file.findAll('movie',{"rated":True})
        for rating in ratings:
            dict_250['rating'] = str(rating['rated'])
        
        release_dates=soup_file.findAll('movie',{"released":True})
        for release in release_dates:
            dict_250['released'] = str(release['released'])
        
        run_times=soup_file.findAll('movie',{"runtime":True})
        for run_time in run_times:
            dict_250['runtime'] = str(run_time['runtime'])
        
        genres=soup_file.findAll('movie',{"genre":True})
        for genre in genres:
            dict_250['genre'] = str(genre['genre'])
        
        directors=soup_file.findAll('movie',{"director":True})
        for director in directors:
            dict_250['director'] = director['director'].encode('utf8')
        
        writers=soup_file.findAll('movie',{"writer":True})
        for writer in writers:
            dict_250['writer'] = writer['writer'].encode('utf8')
        
        actors=soup_file.findAll('movie',{"actors":True})
        for actor in actors:
            dict_250['actors'] = actor['actors'].encode('utf8')
        
        plots=soup_file.findAll('movie',{"plot":True})
        for plot in plots:
            dict_250['plot'] = plot['plot'].encode('utf8')
        
        languages=soup_file.findAll('movie',{"language":True})
        for language in languages:
            dict_250['language'] = str(language['language'])
        
        countrys=soup_file.findAll('movie',{"country":True})
        for country in countrys:
            dict_250['country'] = str(country['country'])
        
        awards=soup_file.findAll('movie',{"awards":True})
        for award in awards:
            dict_250['awards'] = str(award['awards'])
        
        metascores=soup_file.findAll('movie',{"metascore":True})
        for metascore in metascores:
            dict_250['metascore'] = str(metascore['metascore'])
        
        imdbRatings=soup_file.findAll('movie',{"imdbRating":True})
        for imdbRating in imdbRatings:
            dict_250['imdbRating'] = str(imdbRating['imdbRating'])
        
        imdbVotes=soup_file.findAll('movie',{"imdbVotes":True})
        for imdbVote in imdbVotes:
            dict_250['imdbVotes'] = str(imdbVote['imdbVotes'])
        
        imdbIDs=soup_file.findAll('movie',{"imdbID":True})
        for imdbID in imdbIDs:
            dict_250['imdbID'] = str(imdbID['imdbID'])
        
        tomatoMeters=soup_file.findAll('movie',{"tomatoMeter":True})
        for tomatoMeter in tomatoMeters:
            dict_250['tomatoMeter'] = str(tomatoMeter['tomatoMeter'])
    
        tomatoImages=soup_file.findAll('movie',{"tomatoImage":True})
        for tomatoImage in tomatoImages:
            dict_250['tomatoImage'] = str(tomatoImage['tomatoImage'])
                
        tomatoRatings=soup_file.findAll('movie',{"tomatoRating":True})
        for tomatoRating in tomatoRatings:
            dict_250['tomatoRating'] = str(tomatoRating['tomatoRating'])
                
        tomatoReviews=soup_file.findAll('movie',{"tomatoReviews":True})
        for tomatoReview in tomatoReviews:
            dict_250['tomatoReviews'] = str(tomatoReview['tomatoReviews'])
                
        tomatoFreshs=soup_file.findAll('movie',{"tomatoFresh":True})
        for tomatoFresh in tomatoFreshs:
            dict_250['tomatoFresh'] = str(tomatoFresh['tomatoFresh'])
                
        tomatoRottens=soup_file.findAll('movie',{"tomatoRotten":True})
        for tomatoRotten in tomatoRottens:
            dict_250['tomatoRotten'] = str(tomatoRotten['tomatoRotten'])
                
        tomatoConsensuses=soup_file.findAll('movie',{"tomatoConsensus":True})
        for tomatoConsensus in tomatoConsensuses:
            dict_250['tomatoConsensus'] = tomatoConsensus['tomatoConsensus'].encode('utf8')
                
        tomatoUserMeters=soup_file.findAll('movie',{"tomatoUserMeter":True})
        for tomatoUserMeter in tomatoUserMeters:
            dict_250['tomatoUserMeter'] = tomatoUserMeter['tomatoUserMeter'].encode('utf8')
                
        tomatoUserRatings=soup_file.findAll('movie',{"tomatoUserRating":True})
        for tomatoUserRating in tomatoUserRatings:
            dict_250['tomatoUserRating'] = str(tomatoUserRating['tomatoUserRating'])
                
        tomatoUserReviews=soup_file.findAll('movie',{"tomatoUserReviews":True})
        for tomatoUserReview in tomatoUserReviews:
            dict_250['tomatoUserReviews'] = str(tomatoUserReview['tomatoUserReviews'])
                
        DVDs=soup_file.findAll('movie',{"DVD":True})
        for DVD in DVDs:
            dict_250['DVD'] = str(DVD['DVD'])
                
        BoxOffices=soup_file.findAll('movie',{"BoxOffice":True})
        for BoxOffice in BoxOffices:
            dict_250['BoxOffice'] = str(BoxOffice['BoxOffice'])
                
        Productions=soup_file.findAll('movie',{"Production":True})
        for Production in Productions:
            dict_250['Production'] = str(Production['Production'])
        
        list_dict_250.append(dict_250)        
    
    return list_dict_250

top_250_raw = extract_everything(results)


In [116]:
trawdf = pd.DataFrame.from_records(top_250_raw)
trawdf.head()
trawdf.info

<bound method DataFrame.info of     BoxOffice          DVD                       Production  \
0         N/A  27 Jan 1998                Columbia Pictures   
1         N/A  09 Oct 2001               Paramount Pictures   
2         N/A  24 May 2005               Paramount Pictures   
3         N/A  09 Dec 2008  Warner Bros. Pictures/Legendary   
4         N/A  09 Mar 2004               Universal Pictures   
5         N/A  06 Mar 2001             Criterion Collection   
6         N/A  19 May 1998                    Miramax Films   
7         N/A  25 May 2004                  New Line Cinema   
8         N/A  18 Nov 1997                   United Artists   
9         N/A  06 Jun 2000                 20th Century Fox   
10        N/A  06 Aug 2002                  New Line Cinema   
11        N/A  21 Sep 2004            Twentieth Century Fox   
12        N/A  28 Aug 2001               Paramount Pictures   
13        N/A  07 Dec 2010            Warner Bros. Pictures   
14        N/A  26 Aug 2